File: E_Kotlin/Projects/JuegoSnake/app/src/main/java/com/sebss/juegosnake/androidimpl/AndroidAudio.java
package com.sebss.juegosnake.androidimpl;

import java.io.IOException;

import android.app.Activity;
import android.content.res.AssetFileDescriptor;
import android.content.res.AssetManager;
import android.media.AudioManager;
import android.media.SoundPool;

import com.sebss.juegosnake.Audio;
import com.sebss.juegosnake.Musica;
import com.sebss.juegosnake.Sonido;


public class AndroidAudio implements Audio {
    AssetManager assets;
    SoundPool soundPool;

    public AndroidAudio(Activity activity) {
        activity.setVolumeControlStream(AudioManager.STREAM_MUSIC);
        this.assets = activity.getAssets();
        this.soundPool = new SoundPool.Builder()
                .setMaxStreams(20)
                .build();
    }

    @Override
    public Musica nuevaMusica(String filename) {
        try {
            AssetFileDescriptor assetDescriptor = assets.openFd(filename);
            return new AndroidMusica(assetDescriptor);
        } catch (IOException e) {
            throw new RuntimeException("no se ha podido cargar archivo '" + filename + "'");
        }
    }

    @Override
    public Sonido nuevoSonido(String filename) {
        try {
            AssetFileDescriptor assetDescriptor = assets.openFd(filename);
            int soundId = soundPool.load(assetDescriptor, 0);
            return new AndroidSonido(soundPool, soundId);
        } catch (IOException e) {
            throw new RuntimeException("No se ha podido cargar archivo '" + filename + "'");
        }
    }
}

File: E_Kotlin/Projects/AndroidStudio/XML/template_games/androidimpl/AndroidAudio.kt
package com.sebss.pass_word.androidimpl

import android.app.Activity
import android.content.res.AssetManager
import android.media.AudioManager
import android.media.SoundPool
import com.sebss.pass_word.interfaces.Audio
import com.sebss.pass_word.interfaces.Music
import com.sebss.pass_word.interfaces.Sound
import java.io.IOException

class AndroidAudio(activity: Activity) : Audio {
    var assets: AssetManager
    private var soundPool: SoundPool

    init {
        activity.volumeControlStream = AudioManager.STREAM_MUSIC
        assets = activity.assets
        soundPool = SoundPool.Builder()
            .setMaxStreams(20)
            .build()
    }

    override fun newMusic(fileName: String): Music {
        return try {
            val assetDescriptor = assets.openFd(fileName)
            AndroidMusic(assetDescriptor)
        } catch (e: IOException) {
            throw RuntimeException("no se ha podido cargar archivo '$fileName'")
        }
    }

    override fun newSound(fileName: String): Sound {
        return try {
            val assetDescriptor = assets.openFd(fileName)
            val soundId = soundPool.load(assetDescriptor, 0)
            AndroidSound(soundPool, soundId)
        } catch (e: IOException) {
            throw RuntimeException("No se ha podido cargar archivo '$fileName'")
        }
    }
}